In [4]:
# Basic liberaries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# sklearn liberaries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler , OneHotEncoder
# Ignore warnings

import warnings
warnings.filterwarnings('ignore')

In [3]:
# Importing the data
data = pd.read_csv('Housing.csv')
data.head()

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


In [5]:
# Shape of the data
data.shape

(545, 13)

In [7]:
# missing values
data.isnull().sum()

price               0
area                0
bedrooms            0
bathrooms           0
stories             0
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     0
parking             0
prefarea            0
furnishingstatus    0
dtype: int64

In [8]:
# Dependency split

x = data.drop('price',axis=1)
y = data[['price']]

print(x.shape)
print(y.shape)

(545, 12)
(545, 1)


In [9]:
# Train Test Split on x, y

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)
print(x_train.shape,x_test.shape,y_train.shape,y_test.shape, sep='\n')

(381, 12)
(164, 12)
(381, 1)
(164, 1)


In [10]:
# Filtering the categorical columns
cat_cols = [cols for cols in data.columns if data[cols].dtype == 'O']

In [21]:
# OneHotEncoding on train test split data
# fit transform on train data

encoder = OneHotEncoder(sparse_output=False, dtype='int', drop='first')
x_train_encoded = encoder.fit_transform(x_train[cat_cols])
x_train_encoded = pd.DataFrame(x_train_encoded, columns = encoder.get_feature_names_out())
x_train_encoded_df = pd .concat([x_train.drop(cat_cols, axis=1), x_train_encoded],axis=1)
x_train_encoded_df.head()

,area,bedrooms,bathrooms,stories,parking,mainroad_yes,guestroom_yes,basement_yes,hotwaterheating_yes,airconditioning_yes,prefarea_yes,furnishingstatus_semi-furnished,furnishingstatus_unfurnished
126,7160.0,3.0,1.0,1.0,2.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
363,3584.0,2.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
370,4280.0,2.0,1.0,1.0,2.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
31,7000.0,3.0,1.0,4.0,2.0,1.0,1.0,1.0,0.0,1.0,1.0,0.0,1.0
113,9620.0,3.0,1.0,1.0,2.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0


In [23]:
x_train_encoded_df.shape

(491, 13)

# Observation

* if you can see the shape of above train data, the shape has changed from 381 to 491.
* it is because of the index's when we are concatenating both x_train.drop(cat_cols, axis=1), x_train_encoded],axis=1.
* As we are concatenating the new DF is reseting the index's.

In [27]:
#Resetting the index use index=

encoder = OneHotEncoder(sparse_output=False, dtype='int', drop='first')
x_train_encoded = encoder.fit_transform(x_train[cat_cols])
x_train_encoded = pd.DataFrame(x_train_encoded,index=x_train.index, columns = encoder.get_feature_names_out())
x_train_encoded_df = pd .concat([x_train.drop(cat_cols, axis=1), x_train_encoded],axis=1)
x_train_encoded_df.head()

,area,bedrooms,bathrooms,stories,parking,mainroad_yes,guestroom_yes,basement_yes,hotwaterheating_yes,airconditioning_yes,prefarea_yes,furnishingstatus_semi-furnished,furnishingstatus_unfurnished
126,7160,3,1,1,2,1,0,1,0,0,1,0,1
363,3584,2,1,1,0,1,0,0,1,0,0,1,0
370,4280,2,1,1,2,1,0,0,0,1,0,1,0
31,7000,3,1,4,2,1,0,0,0,1,0,1,0
113,9620,3,1,1,2,1,0,1,0,0,1,0,0


In [28]:
x_train_encoded_df.shape

(381, 13)

In [35]:
# test data
#Transform on test

x_test_encoded = encoder.transform(x_test[cat_cols])
x_test_encoded = pd.DataFrame(x_test_encoded, index=x_test.index, columns = encoder.get_feature_names_out())
x_test_encoded_df = pd .concat([x_test.drop(cat_cols, axis=1), x_test_encoded],axis=1)
x_test_encoded_df.head()

,area,bedrooms,bathrooms,stories,parking,mainroad_yes,guestroom_yes,basement_yes,hotwaterheating_yes,airconditioning_yes,prefarea_yes,furnishingstatus_semi-furnished,furnishingstatus_unfurnished
316,5900,4,2,2,1,0,0,1,0,0,0,0,1
77,6500,3,2,3,0,1,0,0,0,1,1,0,0
360,4040,2,1,1,0,1,0,0,0,0,0,1,0
90,5000,3,1,2,0,1,0,0,0,1,0,1,0
493,3960,3,1,1,0,1,0,0,0,0,0,0,0


In [33]:
x_test_encoded_df.shape

(164, 13)

In [37]:
# Scaling

scaler = StandardScaler()

#fit transform on train

x_train_scaled = scaler.fit_transform(x_train_encoded_df)
x_train_scaled_df = pd.DataFrame(x_train_scaled, index=x_train_encoded_df.index, columns=scaler.get_feature_names_out())

#transform on test

x_test_scaled = scaler.transform(x_test_encoded_df)
x_test_scaled_df = pd.DataFrame(x_test_scaled, index=x_test_encoded_df.index, columns=scaler.get_feature_names_out())
